controllers/progressivesync_controller.go

```diff
@@ -341,168 +281,272 @@ func (r *ProgressiveSyncReconciler) getOwnedAppsFromClusters(ctx context.Context
+		log.Error(err, "unabled to update the state map")
+		return syncv1alpha1.StageStatusFailed, err
 	}
 
-	if scheduler.IsStageFailed(apps, stage, pss) {
-		message := fmt.Sprintf("%s stage failed", stage.Name)
-		log.Info(message)
+	// If any adopted app is failed, fail the stage
+	if len(utils.GetAppsByHealthStatusCode(syncedInCurrentStage, health.HealthStatusDegraded)) > 0 {
+		return syncv1alpha1.StageStatusFailed,
+			fmt.Errorf("apps %s health status degraded",
+				utils.GetAppsName(utils.GetAppsByHealthStatusCode(syncedInCurrentStage, health.HealthStatusDegraded)))
+	}
+
+	// To be on the safe side, the controller looks at ALL the progressing Applications selected by the stage selector.
+	// This to avoid having too many Applications in progress triggered by an external process.
+	progressingApps := utils.GetAppsByHealthStatusCode(syncedApps, health.HealthStatusProgressing)
 
-		r.updateStageStatus(ctx, stage.Name, message, syncv1alpha1.PhaseFailed, &ps)
+	// If we reached the maximum number of progressing apps for the stage
+	// then the stage is progressing
+	if len(progressingApps) >= maxTargets {
+		return syncv1alpha1.StageStatusProgressing, nil
+	}
 
-		failed := ps.NewStatusCondition(syncv1alpha1.CompletedCondition, metav1.ConditionFalse, syncv1alpha1.StagesFailedReason, message)
-		apimeta.SetStatusCondition(ps.GetStatusConditions(), failed)
+	outOfSyncApps := utils.GetAppsBySyncStatusCode(selectedApps, argov1alpha1.SyncStatusCodeOutOfSync)
 
-		log.Info("sync failed")
-		return ps, ctrl.Result{Requeue: true, RequeueAfter: RequeueDelayOnError}, nil
+	// If there are no out-of-sync apps then the stage is completed
+	if len(outOfSyncApps) == 0 {
+		return syncv1alpha1.StageStatusCompleted, nil
 	}
 
-	if scheduler.IsStageInProgress(apps, stage, pss) {
-		message := fmt.Sprintf("%s stage in progress", stage.Name)
-		log.Info(message)
+	log.Info("apps summary", "progressingApps", utils.GetAppsName(progressingApps), "syncedInCurrentStage", utils.GetAppsName(syncedInCurrentStage), "OutOfSync", utils.GetAppsName(outOfSyncApps), "stage", stage.Name)
 
-		for _, app := range apps {
-			log.Info("application details", "app", fmt.Sprintf("%s/%s", app.Namespace, app.Name), "sync.status", app.Status.Sync.Status, "health.status", app.Status.Health.Status)
-		}
+	// If there is an external process triggering a sync,
+	// maxParallel - len(progressingApps) might actually be greater than len(outOfSyncApps)
+	// causing the runtime to panic
+	maxSync := maxParallel - len(progressingApps)
+	if maxSync > len(outOfSyncApps) {
+		maxSync = len(outOfSyncApps)
+	}
 
-		r.updateStageStatus(ctx, stage.Name, message, syncv1alpha1.PhaseProgressing, &ps)
+	// If the number of apps synced in the current stage
+	// is equal or greater than the maximum number of targets to sync,
+	// there is nothing else to do.
+	if len(syncedInCurrentStage) >= maxTargets {
+		return syncv1alpha1.StageStatusCompleted, nil
+	}
```

Comment:
perhaps move both completed checks next to each other

Reply:
Done